## **Universidad Nacional de San Antonio Abad del Cusco**
### **Departamento Académico de Ing.Informática**
# **Deep Learning**
## Práctica 08

## **REDES NEURONALES CONVOLUCIONALES (CNN)**

**1.	Importe tensorflow.keras.datasets.mnist como mnist:**

In [1]:
import tensorflow.keras.datasets.mnist as mnist

**2.	Cargue el conjunto de datos mnist usando mnist.load_data() y guarde los resultados en (features_train, label_train), (features_test, label_test):**

In [2]:
(features_train, label_train), (features_test, label_test) = mnist.load_data()

**3.	Imprime el contenido de label_train:**

In [3]:
label_train

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

**4.	Imprime la forma del conjunto de entrenamiento:**

In [4]:
features_train.shape

(60000, 28, 28)

**5.	Imprima la forma del conjunto de prueba:**

In [5]:
features_test.shape

(10000, 28, 28)

**6.	Cambie la forma de los conjuntos de entrenamiento y prueba con las dimensiones (number_observations, 32, 32, 1):**

In [6]:
features_train = features_train.reshape(60000, 32, 32, 1)
features_test = features_test.reshape(10000, 32, 32, 1)

ValueError: cannot reshape array of size 47040000 into shape (60000,32,32,1)

In [ ]:
import matplotlib.pyplot as plt
plt.imshow(x_train[10000])
print("Class: ", y_train[10000])

**7.	Estandarice features_train y features_test dividiéndolos entre 255:**

In [ ]:
features_train = features_train / 255.0
features_test = features_test / 255.0

**8.	Importe numpy como np, tensorflow como tf y layers de tensorflow.keras:**

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers

**9.	Establezca 8 como semilla para numpy y tensorflow usando np.random_seed() y tf.random.set_seed (), respectivamente:**

In [ ]:
np.random.seed(8)
tf.random.set_seed(8)

**10.	Cree una instancia de una clase tf.keras.Sequential () y guárdela en una variable llamada modelo:**

In [ ]:
model = tf.keras.Sequential()

**11.	Cree una instancia de una clase layer.Conv2D () con 64 kernels de forma (3,3), activación = 'relu' y input_shape = (28,28,1), y guárdelo en una variable llamada conv_layer1:**

In [ ]:
conv_layer1 = layers.Conv2D(64, (3,3), activation='relu', input_shape=(28, 28, 1))

**12.	Cree una instancia de una clase layer.Conv2D () con 64 núcleos de forma (3,3) y activación = 'relu' y guárdela en una variable llamada conv_layer2:**

In [ ]:
conv_layer2 = layers.Conv2D(64, (3,3), activation='relu')

**13.	Cree una instancia de la clase layer.Dense () con 128 neuronas, activación = 'relu', y guárdela en una variable llamada fc_layer1:**

In [ ]:
fc_layer1 = layers.Dense(128, activation='relu')

**14.	Cree una instancia de la clase layer.Dense () con 10 neuronas, activación = 'softmax', y guárdela en una variable llamada fc_layer2:**

In [ ]:
fc_layer2 = layers.Dense(10, activation='softmax')

**15.	Agregue las cuatro capas que acaba de definir al modelo usando .add (), agregue una capa MaxPooling2D () de tamaño (2,2) entre cada una de las capas de convolución y agregue una capa Flatten() antes de la primera capa completamente conectada para aplanar los mapas de características:**

In [ ]:
model.add(conv_layer1)
model.add(layers.MaxPooling2D(2, 2))
model.add(conv_layer2)
model.add(layers.MaxPooling2D(2, 2))
model.add(layers.Flatten())
model.add(fc_layer1)
model.add(fc_layer2)

**16.	Cree una instancia de una clase tf.keras.optimizers.Adam () con 0.001 como tasa de aprendizaje y guárdela en una variable llamada optimizer:**

In [ ]:
optimizer = tf.keras.optimizers.Adam(0.001)

**17.	Compile la red neuronal usando .compile() con loss = 'sparse_ categorical_crossentropy', optimizer = optimizer, metrics = ['precision']:**

In [ ]:
model.compile(loss='sparse_categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

**18.	Imprime el resumen del modelo:**

In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 26, 26, 64)        640       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 64)       0         
 )                                                               
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        36928     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 flatten (Flatten)           (None, 1600)              0         
                                                                 
 dense (Dense)               (None, 128)               2

Capa de red  | Entrada    | Filtro      |Stride     |Produce    |Fórmula  |   Parámetros
:---         | :---:      | :---:        | :---:      | :---:     | ---:     | ---:      |   
Input   | 28x28x1  |      |      | 28x28x1 | |0
Conv2d|28x28x1|  3x3x64  |  1   | 26x26x64 | 3x3x1x64+64| 640
maxpooling|26x26x64|2x2|2|13x13x64| |0
Conv2d|13x13x64|3x3x64|1|11x11x64|3x3x64x64+64|36928
maxpooling|11x11x64|2x2|2|5x5x64| | 0
flatten|5x5x64| | |1600| |0
dense|1600| | | 128|1600x128+128|204928
dense2|128| | | 10|128*10+10|1290
TOTAL| | | | | | 243786


**19.	Ajuste las redes neuronales con el conjunto de entrenamiento y especifique epochs = 5, validation_split = 0.2 y verbose = 2:**


si batch_size= None, el nro de muestras es 32 en cada pasada
por ejemplo el nro de datos es 60000, pero 20% se utilizan para validación, entonces el nro total de muestras para entrenamiento es 48000
por consiguiente el nro de pasadas en cada iteración es 1500

In [ ]:
#model.fit(features_train, label_train,batch_size=64, epochs=5, validation_split=0.2, verbose=1)
model.fit(features_train, label_train,batch_size=64, epochs=5, verbose=1)

Epoch 1/5
938/938 [==============================] - 88s 93ms/step - loss: 0.1496 - accuracy: 0.9551
Epoch 2/5
938/938 [==============================] - 91s 97ms/step - loss: 0.0429 - accuracy: 0.9865
Epoch 3/5
938/938 [==============================] - 86s 92ms/step - loss: 0.0305 - accuracy: 0.9905
Epoch 4/5
938/938 [==============================] - 85s 90ms/step - loss: 0.0235 - accuracy: 0.9924
Epoch 5/5
938/938 [==============================] - 86s 92ms/step - loss: 0.0165 - accuracy: 0.9949


**20.	Evaluemos el rendimiento del modelo en el conjunto de pruebas:**

In [ ]:
model.evaluate(features_test, label_test)

313/313 [==============================] - 5s 15ms/step - loss: 0.0309 - accuracy: 0.9908


[0.03091258928179741, 0.9908000230789185]